Code/Framework/AzCore/AzCore/DOM/DomPatch.cpp

```diff
@@ -0,0 +1,959 @@
+            {
+                return AZ::Failure(pathLoad.TakeError());
+            }
+            auto valueLoad = loadField("value");
+            if (!valueLoad.IsSuccess())
+            {
+                return AZ::Failure(valueLoad.TakeError());
+            }
+
+            return AZ::Success(PatchOperation::TestOperation(Path(pathLoad.GetValue().GetString()), valueLoad.TakeValue()));
+        }
+        else
+        {
+            return AZ::Failure<AZStd::string>("PatchOperation failed to create DOM representation: invalid \"op\" specified");
+        }
+    }
+
+    AZ::Outcome<PatchOperation, AZStd::string> PatchOperation::GetInverse(Value stateBeforeApplication) const
+    {
+        switch (m_type)
+        {
+        case Type::Add:
+            {
+                // Add -> Replace (if value already existed in an object) otherwise
+                // Add -> Remove
+                if (m_domPath.Size() > 0 && m_domPath[m_domPath.Size() - 1].IsKey())
+                {
+                    const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                    if (existingValue != nullptr)
+                    {
+                        return AZ::Success(PatchOperation::ReplaceOperation(m_domPath, *existingValue));
+                    }
+                }
+                return AZ::Success(PatchOperation::RemoveOperation(m_domPath));
+            }
+        case Type::Remove:
+            {
+                // Remove -> Add
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    return AZ::Failure(
+                        AZStd::string::format("Unable to invert DOM remove patch, source path not found: %s", m_domPath.ToString().data()));
+                }
+                return AZ::Success(PatchOperation::AddOperation(m_domPath, *existingValue));
+            }
+        case Type::Replace:
+            {
+                // Replace -> Replace (with old value)
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    return AZ::Failure(AZStd::string::format(
+                        "Unable to invert DOM replace patch, source path not found: %s", m_domPath.ToString().data()));
+                }
+                return AZ::Success(PatchOperation::ReplaceOperation(m_domPath, *existingValue));
+            }
+        case Type::Copy:
+            {
+                // Copy -> Replace (with old value)
+                const Value* existingValue = stateBeforeApplication.FindChild(m_domPath);
+                if (existingValue == nullptr)
+                {
+                    return AZ::Failure(
+                        AZStd::string::format("Unable to invert DOM copy patch, source path not found: %s", m_domPath.ToString().data()));
```

Comment:
Alternatively this can follow JsonPatch error output, where we use a [fixed_string](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/Json/JsonMerger.cpp#L19) for errors

Reply:
I like the idea of using fixed strings in the generic DOM API and will take a note on that for a follow-up. For now I've made these all use `AppendString` and added a fixed_string compatible AppendString API to `Path` in case we want to go the fixed string path later.